File: 945_minimum_increment_to_make_array_unique/minIncrementForUniqueMap.ts
function minIncrementForUnique(nums: number[]): number {
  const root = new Map();

  function find(x: number): number {
    const value = root.has(x) ? find(root.get(x) + 1) : x;
    root.set(x, value);
    return root.get(x);
  }

  return nums.reduce((a, v) => (a += find(v) - v), 0);
}

File: 945_minimum_increment_to_make_array_unique/minIncrementForUniqueMap.py
from ast import List


class Solution:
    def minIncrementForUnique(self, nums: List[int]) -> int:
        root = {}

        def find(x):
            value = find(root[x] + 1) if x in root else x
            root[x] = value

            return root[x]

        return sum(find(num) - num for num in nums)